crates/neon-runtime/src/napi/tsfn.rs

```diff
@@ -167,6 +195,14 @@ impl<T: Send + 'static> ThreadsafeFunction<T> {
 
 impl<T> Drop for ThreadsafeFunction<T> {
     fn drop(&mut self) {
+        let is_finalized = self.is_finalized.lock().unwrap();
+
+        // tsfn was already finalized by `Environment::CleanupHandles()` in
+        // Node.js
+        if *is_finalized {
+            return;
+        }
```

Comment:
Don't you need to drop the lock before calling `napi::release_threadsafe_function`?

Reply:
Nope, actually I don't the idea here is that we want `finalize_cb` to block until `release_threadsafe_function` completes, because otherwise `finalize_cb` might finish first and delete the tsfn in C++.

Reply:
Mutex isn't re-entrant, though, so in the *non*-early-exit case you'll lock, call `release_threadsafe_function`, and fail when it's time to lock again.

Reply:
`release_threadsafe_function` doesn't call `finalize_cb` directly. It schedules an event on the loop thread and executes `finalize_cb` in another tick.